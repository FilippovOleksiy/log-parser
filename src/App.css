.App {
  text-align: center;
  background-color: #282c34;
  min-height: 100vh;
  min-width: 300px;
  display: flex;
  font-size: calc(10px + 2vmin);
  color: white;
  position: relative;
}
.App-content-section {
  margin: 0 0 0 10vw;
}

.App-logo-container {
  width: 10vw;
  position: absolute;
  right: 0;
  overflow: hidden;
}

.App-logo {
  width: 10vw;
  pointer-events: none;
}

.Choose-file {
  display: flex;
  margin: 2vh 0 2vh 0;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
