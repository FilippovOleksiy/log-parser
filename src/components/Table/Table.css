.Table {
  display: flex;
  font-size: 1.3vw;
  flex-direction: column;
  margin: 4vh 0 0 0;
}

.Table-header {
  color: #61dafb;
  padding: 0 0 2vh 0;
}

.Table-row,
.Table-header {
  display: flex;
  flex-direction: row;
}

.Table-row > div {
  display: flex;
  width: 50%;
}

.Table-header > div {
  display: flex;
  width: 40%;
  margin: 0 10% 0 0;
}

.Table-body > .Table-row:nth-child(odd) {
  background-color: #354566;
}

.Sort--selected {
  border-bottom: 1px solid;
}

.Sort {
  cursor: pointer;
}

.Sort-direction {
  margin: 0 0 0 1vw;
}
