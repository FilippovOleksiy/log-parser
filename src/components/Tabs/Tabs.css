.Tabs {
  display: flex;
  font-size: 1.5vw;
}

.Tab-item {
  padding: 5px;
  margin: 0 20px 0 0;
  color: #61dafb;
  padding: 1vw;
  border: 1px solid transparent;
  border-radius: 10px;
  cursor: pointer;
}

.Tab-item--selected {
  padding: 1vw;
  border: 1px solid;
  border-radius: 10px;
}
