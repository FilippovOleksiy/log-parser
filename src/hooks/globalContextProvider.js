import React, { createContext, useContext, useState, useMemo } from "react";

const ReactGlobalContext = createContext(null);

function GlobalContextProvider({ children }) {
  const [context, setContext] = useState({});
  const value = useMemo(() => ({ context, setContext }), [context, setContext]);

  return (
    <ReactGlobalContext.Provider value={value}>
      {children}
    </ReactGlobalContext.Provider>
  );
}

const useGlobalContext = () => {
  const context = useContext(ReactGlobalContext);
  if (context === undefined) {
    throw new Error(
      "useGlobalContext must be used within a ReactGlobalContext"
    );
  }
  return context;
};

export { GlobalContextProvider, useGlobalContext };
