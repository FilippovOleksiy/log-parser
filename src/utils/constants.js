export const tabsList = ["Total webpages views", "Total uniq webpages views"];
export const viewsSortIndicators = ["N..n", "n..N"];
export const pageNameSortIndicators = ["A..Z", "Z..A"];
