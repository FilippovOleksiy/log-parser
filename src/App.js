import { useEffect, useMemo, useState } from "react";

import logo from "./logo.svg";
import "./App.css";

import { useGlobalContext } from "./hooks/globalContextProvider";
import { fetchUniqWebpages, fetchWebpages } from "./utils/index";
import { tabsList } from "./utils/constants";

import Tabs from "./components/Tabs";
import Table from "./components/Table";

function App() {
  const [fileContent, setFileContent] = useState(null);
  const [webpages, setWebpages] = useState(null);

  const {
    context: { selectedTab },
  } = useGlobalContext();

  const uniqWebpagesMemo = useMemo(
    () => (fileContent && Object.entries(fetchUniqWebpages(fileContent))) || [],
    [fileContent]
  );

  const webpagesMemo = useMemo(
    () =>
      (uniqWebpagesMemo && Object.entries(fetchWebpages(uniqWebpagesMemo))) ||
      [],
    [uniqWebpagesMemo]
  );

  useEffect(() => {
    if (selectedTab === tabsList[0]) {
      setWebpages(webpagesMemo);
    } else {
      setWebpages(uniqWebpagesMemo);
    }
  }, [uniqWebpagesMemo, webpagesMemo, selectedTab]);

  const onChange = (e) => {
    const reader = new FileReader();
    reader.onload = async (event) => setFileContent(event.target.result);
    reader.readAsText(e.target.files[0]);
  };

  return (
    <div className="App">
      <div className="App-content-section">
        <div className="Choose-file">
          <input type="file" accept=".log" onChange={onChange} />
        </div>
        {!!webpages?.length && <Tabs />}
        {!!webpages?.length && <Table webpages={webpages} />}
      </div>
      <div>
        <div className="App-logo-container">
          <img src={logo} className="App-logo" alt="logo" />
        </div>
      </div>
    </div>
  );
}

export default App;
