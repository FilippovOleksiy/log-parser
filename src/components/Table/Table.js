import React, { useCallback, useEffect, useMemo, useState } from "react";

import "./Table.css";

import { sortByPageName, sortByViews } from "../../utils/index";
import {
  viewsSortIndicators,
  pageNameSortIndicators,
} from "../../utils/constants";

function Table({ webpages }) {
  const [viewsSortDirections, setViewsSortDirections] = useState(null);
  const [pageNameSortDirections, setPageNameSortDirections] = useState(
    pageNameSortIndicators
  );
  const [sortedList, setSortedList] = useState([]);

  const sortedWebpages = useMemo(() => {
    if (pageNameSortDirections) {
      return sortByPageName(webpages);
    } else {
      return sortByViews(webpages);
    }
  }, [webpages, pageNameSortDirections]);

  const sortClickHandler = useCallback(
    (sortDirection) => {
      if (sortDirection === null) {
        if (sortDirection === pageNameSortDirections) {
          setViewsSortDirections(null);
          setPageNameSortDirections(pageNameSortIndicators);
        } else {
          setViewsSortDirections(viewsSortIndicators);
          setPageNameSortDirections(null);
        }
        setSortedList(sortedWebpages);
      } else {
        if (sortDirection === pageNameSortDirections) {
          setPageNameSortDirections((value) => value.slice().reverse());
        } else {
          setViewsSortDirections((value) => value.slice().reverse());
        }

        setSortedList((list) => list.slice().reverse());
      }
    },
    [pageNameSortDirections, sortedWebpages]
  );

  useEffect(() => {
    setSortedList(sortedWebpages);
    setViewsSortDirections(null);
    setPageNameSortDirections(pageNameSortIndicators);
  }, [webpages, sortedWebpages]);

  return (
    <div className="Table">
      <div className="Table-header">
        <div
          className={`Sort ${pageNameSortDirections && "Sort--selected"}`}
          onClick={() => sortClickHandler(pageNameSortDirections)}
        >
          Webpage name
          {pageNameSortDirections && (
            <span className="Sort-direction">
              ({pageNameSortDirections[0]})
            </span>
          )}
        </div>
        <div
          className={`Sort ${viewsSortDirections && "Sort--selected"}`}
          onClick={() => sortClickHandler(viewsSortDirections)}
        >
          Views
          {viewsSortDirections && (
            <span className="Sort-direction">({viewsSortDirections[0]})</span>
          )}
        </div>
      </div>

      <div className="Table-body">
        {sortedList.map(([key, value]) => (
          <div key={key} className={"Table-row"}>
            <div>{key}</div>
            <div>{value}</div>
          </div>
        ))}
      </div>
    </div>
  );
}

export default Table;
