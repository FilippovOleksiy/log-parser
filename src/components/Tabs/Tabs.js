import React, { useCallback, useEffect } from "react";

import "./Tabs.css";

import { useGlobalContext } from "../../hooks/globalContextProvider";
import { tabsList } from "../../utils/constants";

function Tabs() {
  const {
    context: { selectedTab },
    setContext,
  } = useGlobalContext();

  const onClick = useCallback(
    (selectedItem) => {
      setContext((cont) => ({ ...cont, selectedTab: selectedItem }));
    },
    [setContext]
  );

  useEffect(() => {
    if (selectedTab === undefined) {
      setContext((cont) => ({ ...cont, selectedTab: tabsList[0] }));
    }
  }, [setContext, selectedTab]);

  return (
    <div className="Tabs">
      {tabsList.map((item) => (
        <div
          key={item}
          className={`Tab-item ${selectedTab === item && "Tab-item--selected"}`}
          onClick={() => {
            onClick(item);
          }}
        >
          {item}
        </div>
      ))}
    </div>
  );
}

export default Tabs;
