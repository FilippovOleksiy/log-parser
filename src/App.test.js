import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";

import App from "./App";
import Tabs from "./components/Tabs";
import Table from "./components/Table";
import { GlobalContextProvider } from "./hooks/globalContextProvider";
import { tabsList } from "./utils/constants";
import { fetchUniqWebpages } from "./utils/index";

const pages = `
/help_page 126.318.035.038
/help_page/2 126.318.035.038
/help_page/1 126.318.035.038
/contact 184.123.665.067
/home 184.123.665.067
/about/2 444.701.448.104
`;

describe("App", () => {
  it("snapshot should match", () => {
    const tree = renderer
      .create(
        <GlobalContextProvider>
          <App />
        </GlobalContextProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Tabs", () => {
  it("snapshot should match", () => {
    const tree = renderer
      .create(
        <GlobalContextProvider>
          <Tabs />
        </GlobalContextProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("first tab should be selected", () => {
    const rendered = render(
      <GlobalContextProvider>
        <Tabs />
      </GlobalContextProvider>
    );
    userEvent.click(screen.getByText(tabsList[0]));
    const div = rendered.container.querySelector("div.Tab-item--selected");

    expect(div.innerHTML).toEqual(tabsList[0]);
  });

  it("second tab should be selected", () => {
    const rendered = render(
      <GlobalContextProvider>
        <Tabs />
      </GlobalContextProvider>
    );
    userEvent.click(screen.getByText(tabsList[1]));
    const div = rendered.container.querySelector("div.Tab-item--selected");

    expect(div.innerHTML).toEqual(tabsList[1]);
  });
});

describe("Table", () => {
  it("snapshot should match", () => {
    const webpages = fetchUniqWebpages(pages);
    const tree = renderer
      .create(
        <GlobalContextProvider>
          <Table webpages={Object.entries(webpages)} />
        </GlobalContextProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("sort by Views ASC", () => {
    const webpages = fetchUniqWebpages(pages);
    const rendered = render(
      <GlobalContextProvider>
        <Table webpages={Object.entries(webpages)} />
      </GlobalContextProvider>
    );
    userEvent.click(screen.getByText("Views"));
    const div = rendered.container.querySelector("div.Sort--selected");

    expect(div.innerHTML).toEqual(
      'Views<span class="Sort-direction">(N..n)</span>'
    );
  });

  it("sort by Views DESC", () => {
    const webpages = fetchUniqWebpages(pages);
    const rendered = render(
      <GlobalContextProvider>
        <Table webpages={Object.entries(webpages)} />
      </GlobalContextProvider>
    );
    userEvent.click(screen.getByText("Views"));
    userEvent.click(screen.getByText("Views"));
    const div = rendered.container.querySelector("div.Sort--selected");

    expect(div.innerHTML).toEqual(
      'Views<span class="Sort-direction">(n..N)</span>'
    );
  });

  it("sort by Webpage name DESC", () => {
    const webpages = fetchUniqWebpages(pages);
    const rendered = render(
      <GlobalContextProvider>
        <Table webpages={Object.entries(webpages)} />
      </GlobalContextProvider>
    );
    userEvent.click(screen.getByText("Webpage name"));
    const div = rendered.container.querySelector("div.Sort--selected");

    expect(div.innerHTML).toEqual(
      'Webpage name<span class="Sort-direction">(Z..A)</span>'
    );
  });

  it("sort by Webpage name ASC", () => {
    const webpages = fetchUniqWebpages(pages);
    const rendered = render(
      <GlobalContextProvider>
        <Table webpages={Object.entries(webpages)} />
      </GlobalContextProvider>
    );
    userEvent.click(screen.getByText("Webpage name"));
    userEvent.click(screen.getByText("Webpage name"));
    const div = rendered.container.querySelector("div.Sort--selected");

    expect(div.innerHTML).toEqual(
      'Webpage name<span class="Sort-direction">(A..Z)</span>'
    );
  });

  it("change sort parameter", () => {
    const webpages = fetchUniqWebpages(pages);
    const rendered = render(
      <GlobalContextProvider>
        <Table webpages={Object.entries(webpages)} />
      </GlobalContextProvider>
    );
    userEvent.click(screen.getByText("Webpage name"));
    userEvent.click(screen.getByText("Views"));
    const div = rendered.container.querySelector("div.Sort--selected");

    expect(div.innerHTML).toEqual(
      'Views<span class="Sort-direction">(N..n)</span>'
    );
  });
});
