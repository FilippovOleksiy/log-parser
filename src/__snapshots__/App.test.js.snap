// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot should match 1`] = `
<div
  className="App"
>
  <div
    className="App-content-section"
  >
    <div
      className="Choose-file"
    >
      <input
        accept=".log"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
  <div>
    <div
      className="App-logo-container"
    >
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
    </div>
  </div>
</div>
`;

exports[`Table snapshot should match 1`] = `
<div
  className="Table"
>
  <div
    className="Table-header"
  >
    <div
      className="Sort Sort--selected"
      onClick={[Function]}
    >
      Webpage name
      <span
        className="Sort-direction"
      >
        (
        A..Z
        )
      </span>
    </div>
    <div
      className="Sort null"
      onClick={[Function]}
    >
      Views
    </div>
  </div>
  <div
    className="Table-body"
  />
</div>
`;

exports[`Tabs snapshot should match 1`] = `
<div
  className="Tabs"
>
  <div
    className="Tab-item Tab-item--selected"
    onClick={[Function]}
  >
    Total webpages views
  </div>
  <div
    className="Tab-item false"
    onClick={[Function]}
  >
    Total uniq webpages views
  </div>
</div>
`;

exports[`renders App correctly 1`] = `
<div
  className="App"
>
  <div
    className="App-content-section"
  >
    <div
      className="Choose-file"
    >
      <input
        accept=".log"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
  <div>
    <div
      className="App-logo-container"
    >
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
    </div>
  </div>
</div>
`;

exports[`renders Tabs correctly 1`] = `
<div
  className="Tabs"
>
  <div
    className="Tab-item Tab-item--selected"
    onClick={[Function]}
  >
    Total webpages views
  </div>
  <div
    className="Tab-item false"
    onClick={[Function]}
  >
    Total uniq webpages views
  </div>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="App"
>
  <div
    className="App-content-section"
  >
    <div
      className="Choose-file"
    >
      <input
        accept=".log"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
  <div>
    <div
      className="App-logo-container"
    >
      <img
        alt="logo"
        className="App-logo"
        src="logo.svg"
      />
    </div>
  </div>
</div>
`;
